### ISSUE TITLE
[Issue]: Automatic invoice payment including manual invoice and/or manual credit fails to update invoice to paid in Mirakl(webhook)

### ISSUE THREAD

**Issue by lsfothergill** [paypal/mirakl-hyperwallet-connector#102]
### Describe the Issue

When the automatic invoice created by Mirakl includes a manual invoice or a manual credit in the total & is sent to Hyperwallet to create the payments (to seller & RCL) when the payment goes to complete in Hyperwallet and the webhook runs to update the invoice to paid in Mirakl it is erroring out due to a mismatch in the amounts. 

Reason: [[400] The amount is incorrect for the accounting document with number '230265', actual: '405.71', expected: '374.36']

NOTE: In this test the automatic invoice had both a manual invoice and manual credit but the same issue happens if an automatic invoice has a manual invoice OR an automatic invoice has a manual credit.  examples can be provided in needed. 

### Environment

Test/UAT

4.6.0 Version

_No response_

### Expected Behavior

The invoice in Mirakl should get set to paid as the payments were created/completed in hyperwallet

### Actual Behavior

The automatic invoice in Mirakl remains pending (as does the manual invoice and/or manual credit). 

### Steps to Reproduce

In the https://reitmansca3-dev.mirakl.net/ environment 
PE-QA Seller 5B was used to run a test where there were payable order lines as well as a manual credit and a manual invoice.
When the billing cycle ran automatic invoice 230265 was created. 
![image](https://user-images.githubusercontent.com/117089103/199013827-545604fb-55a3-4b9f-9ce4-1398215405bd.png)
Invoice 230265 includes the amounts from the manual invoice (230260) & manual credit (230259) so that invoice shows an amount to be transferred of $405.71 – yet in Mirakl screen shot above those amounts are not included in the Payable amount of 230265
![image](https://user-images.githubusercontent.com/117089103/199013902-8ee7b409-b442-43e0-81d7-88abb32f0176.png)
Only automatic invoice 230265 amount is sent to Hyperwallet 
![image](https://user-images.githubusercontent.com/117089103/199014019-b25aa47f-3933-4ec0-92b0-9b6e5b92e808.png)
And once completed it fails when webhook running to update Invoice (or invoices?) to paid.
Due to the actual amount of $405.71 being more than the expected amount of $374.36. (error from log below)

### Pre-conditions

Before the billing cycle runs in order to have an amount to be paid out besides the manual invoice & manual credit RCL Market orders must be created / accepted / shipped for PE-QA Seller 5B. 
The manual invoice and the manual credit must also be created.
The billing cycle must run. 

### Relevant log output

```shell
app_1  | 24-10-2022 17:26:33.189 [http-nio-8080-exec-1] ERROR {"id":"wbh-700b983c-8dd3-44f2-b4c9-9484c78507e6","subtype":"PAYMENTS.UPDATED.STATUS.COMPLETED","targetToken":"pmt-53707f0d-a55f-4816-98f7-fd210ec3a277","clientPaymentId":"230265","type":"Notification"} com.paypal.infrastructure.listeners.AbstractNotificationListener - Notification [wbh-700b983c-8dd3-44f2-b4c9-9484c78507e6] could not be processed. Reason: [[400] The amount is incorrect for the accounting document with number '230265', actual: '405.71', expected: '374.36']
app_1  | 24-10-2022 17:26:33.190 [http-nio-8080-exec-1] ERROR {"id":"wbh-700b983c-8dd3-44f2-b4c9-9484c78507e6","subtype":"PAYMENTS.UPDATED.STATUS.COMPLETED","targetToken":"pmt-53707f0d-a55f-4816-98f7-fd210ec3a277","clientPaymentId":"230265","type":"Notification"} com.paypal.infrastructure.listeners.AbstractNotificationListener - Failed notification with token [wbh-700b983c-8dd3-44f2-b4c9-9484c78507e6] will be stored to be restarted in the future.
```

**Comment by max6001** [comment]
Hi @lsfothergill,

Sorry for the delay but we had to catch-up with Mirakl to understand what's going on in this use case.
We are a bit surprised to see such behaviour and even more to hear it is "normal" knowing we have several marketplaces live with the connector and we would have expected them to run into that issue by now.

Our point of contact in Mirakl is going to investigate and asked, in order to help the investigation, if you could provide the **API result/body of the IV01 API call returning the automatic invoice**. Feel free to reach out to your Mirakl point of contact to get any assistance pulling such data.

Please attach the log to your response on this GitHub ticket as soon as you have it.

Best,

**Comment by lsfothergill** [comment]
IV01 response invoice 230265 (json content)

{
    "invoices": [
        {
            "currency_iso_code": "CAD",
            "date_created": "2022-10-21T04:00:20Z",
            "details": [
                {
                    "amount_excl_taxes": 126.12,
                    "description": "Commissions of completed orders from 10/20/2022 to 10/21/2022",
                    "quantity": 1,
                    "taxes": [
                        {
                            "amount": 6.32,
                            "code": "GST"
                        },
                        {
                            "amount": 12.58,
                            "code": "QST"
                        }
                    ]
                }
            ],
            "due_date": "2022-11-20T05:00:00Z",
            "end_time": "2022-10-21T04:00:20Z",
            "id": "ce13cffd-12b6-4bec-a840-a9642d7cd301",
            "invoice_id": 230265,
            "issue_date": "2022-10-21T04:00:20Z",
            "issuing_user": {
                "information": "Automatic action",
                "type": "BATCH"
            },
            "payment": {
                "reference": null,
                "state": "PENDING",
                "transaction_date": null
            },
            "payment_info": {
                "@type": "CANADIAN",
                "bank_account_number": "1234560005",
                "bank_city": "Montréal",
                "bank_name": "QA Seller 5 Bank Français",
                "bank_street": "1173 St. Catherine St. W.",
                "bank_zip": "H3B 1K4",
                "institution_number": "001",
                "owner": "QA Seller 5 QST & GST",
                "swift_code": null,
                "transit_number": "21211"
            },
            "shop_address": {
                "city": "Montréal",
                "civility": "Miss",
                "country": "CAN",
                "firstname": "Linda",
                "lastname": "Fothergill",
                "phone": "1-866-734-2686",
                "state": "QC",
                "street1": "1171 St. Catherine St. W.",
                "zip_code": "H3B 1K4"
            },
            "shop_corporate_name": "QA Seller 5",
            "shop_id": 2032,
            "shop_model": "MARKETPLACE",
            "shop_name": "PE-QA Seller 5B QST & GST Fr",
            "shop_tax_number": "1234567895",
            "start_time": "2022-10-20T04:00:27Z",
            "state": "COMPLETE",
            "summary": {
                "amount_transferred": 405.71,
                "amount_transferred_to_operator": 178.58,
                "total_commissions_excl_tax": -126.12,
                "total_commissions_incl_tax": -145.02,
                "total_non_payable_orders_excl_tax": 0.00,
                "total_non_payable_orders_incl_tax": 0.00,
                "total_operator_paid_shipping_charges_excl_tax": 0.00,
                "total_operator_paid_shipping_charges_incl_tax": 0.00,
                "total_operator_remitted_taxes": -64.91,
                "total_other_credits_excl_tax": 39.98,
                "total_other_credits_incl_tax": 39.98,
                "total_other_invoices_excl_tax": -7.50,
                "total_other_invoices_incl_tax": -8.63,
                "total_payable_orders_excl_tax": 519.38,
                "total_payable_orders_incl_tax": 584.29,
                "total_purchase_taxes": 0.00,
                "total_refund_commissions_excl_tax": 0.00,
                "total_refund_commissions_incl_tax": 0.00,
                "total_refund_non_payable_orders_excl_tax": 0.00,
                "total_refund_non_payable_orders_incl_tax": 0.00,
                "total_refund_operator_paid_shipping_charges_excl_tax": 0.00,
                "total_refund_operator_paid_shipping_charges_incl_tax": 0.00,
                "total_refund_operator_remitted_taxes": 0.00,
                "total_refund_orders_excl_tax": 0.00,
                "total_refund_orders_incl_tax": 0.00,
                "total_refund_payable_orders_excl_tax": 0.00,
                "total_refund_payable_orders_incl_tax": 0.00,
                "total_refund_purchase_taxes": 0.00,
                "total_subscription_excl_tax": 0.00,
                "total_subscription_incl_tax": 0.00
            },
            "total_amount_excl_taxes": 126.12,
            "total_amount_incl_taxes": 145.02,
            "total_charged_amount": 145.02,
            "total_taxes": [
                {
                    "amount": 6.32,
                    "code": "GST"
                },
                {
                    "amount": 12.58,
                    "code": "QST"
                }
            ],
            "type": "AUTO_INVOICE"
        }
    ],
    "total_count": 1
}

**Comment by max6001** [comment]
Thanks @lsfothergill for your reactivity.
I've sent your response to my Mirakl point of contact.

I'll get back to you asap on this ticket once I'll know more about the outcome.

Best,

**Comment by max6001** [comment]
Hi @lsfothergill,

Mirakl have suggested to re-calculate the automatic invoice amount when processing the notification from HW, in order to take into account the manual invoice and manual credit notes. Unfortunately this isn't something we will do within the connector because the philosophy of the connector is simply to pass over data from Mirakl to HW and vice versa. We do not want the connector to handle any logic, especially when it touches the payment part.

We have suggested two possible solutions to Mirakl:
- If Mirakl were able to provide the value in the IV01 API response, we could cache that value in the connector and use it when we receive the HW notification to update Mirakl.
- Mirakl don't seem to allow partial invoice payment so we beleive it doesn't make any sense to make mandatory the value of the invoice when updating its status. We've recommended to make the amount optionnal when calling the Mirakl API so we wouldn't have to bother with that current issue.

I'm closing this issue until we get any update from Mirakl but don't hesitate to re-open it if you want to add anything.

Best,
